### ISSUE TITLE
Update class properties and decorator interop

### ISSUE THREAD

**Issue by sebmck** [babel/babel#1277]
Descriptors **should not** be stored on the prototype prior to initialization.

**Comment by bucaran** [comment]
I am trying to come up with a solution, but I can't find the template that generates the class.

I think `helper-create-decorated-class` could be tweaked with:

```js
            if (descriptor.initializer) {
                descriptor.value = descriptor.initializer()
            }
```

What do you think?

**Comment by sebmck** [comment]
@bucaran The descriptors shouldn't be stored on the prototype in the first place and should be defined in the constructor.

**Comment by bucaran** [comment]
:+1: 

So what is going to happen to `_createDecoratedClass` ?

**Comment by sebmck** [comment]
It's going to stay since it has nothing to do with property initializers. The property initializer descriptors just wont be included in the list of descriptors and they'll be moved to the class constructor.

**Comment by bucaran** [comment]
Using the same example from #1274 I was looking at the generated class here:

```js
var Person = (function () {
  var _instanceInitializers = {};

  function Person() {
    _classCallCheck(this, Person);

    this.born = _instanceInitializers.born.call(this);
  }

  _createDecoratedClass(Person, [{
    key: "born",
    decorators: [readonly],
    initializer: function () {
      return new Date();
    },
    enumerable: true
  }], null, _instanceInitializers);

  return Person;
})();
```

**Comment by sebmck** [comment]
`this.born = ` would be replaced with `Object.defineProperty(this, "born", ...`

**Comment by bucaran** [comment]
Of course.

**Comment by bucaran** [comment]
If the generated code was

```js
var Person = (function() {
    var _instanceInitializers = {};

    function Person() {
        _classCallCheck(this, Person);

        Object.defineProperty(this, 'born', {
            enumerable: false,
            configurable: false,
            writable: false,
            value: _instanceInitializers.born.call(this)
        });
    }

    _createDecoratedClass(Person, [{
        key: "born",
        decorators: [readonly],
        initializer: function() {
            return new Date();
        },
        enumerable: true
    }], null, _instanceInitializers);

    return Person;
})();
```

It does work, but I am not sure how to modify the class in `classes.js`. Any tips

**Comment by theadam** [comment]
You'll also need to make sure the constructed instance gets the appropriate property descriptor applied to the field.  With this code, Born is not enumerable, but it probably should be.

```javascript
        Object.defineProperty(this, 'born', {
            enumerable: false,
            configurable: false,
            writable: false,
            value: _instanceInitializers.born.call(this)
        });
```

Instead of that, you might want something like this:

```javascript
    var descriptor = Object.getOwnPropertyDescriptor(Person.prototype, 'born');
    if(descriptor.get === undefined && descriptor.set === undefined) {
        descriptor.value =  _instanceInitializers.born.call(this);
    }
    Object.defineProperty(this, 'born', descriptor);
```

Notice, to add "value" to the descriptor, you should probably check that "get" and "set" don't exist on it already.  In this case, I would assume the "get" function on the descriptor would trump the initializer.

**Comment by sebmck** [comment]
Released as of 5.1.12. Sorry @bucaran, I was in the middle of typing out an explanation and in order to actually wrap my head around how to do it I actually had to do it myself :disappointed: See #1347 for other ways in which you can help out. Thanks!

**Comment by bucaran** [comment]
@sebmck Thanks!
